### ISSUE TITLE
`no-function-expression` and generic functions in TSX

### ISSUE THREAD

**Issue by astorije** [Microsoft/tslint-microsoft-contrib#518]
While the following is supported by the language:

```tsx
function foo<T>(x: T): T { return x; }
```

The following is not:

```tsx
const foo = <T>(x: T) => x; // ERROR : unclosed `T` tag
```

`no-function-expression` encourages to use arrow functions, which can lead to the following to make the linter pass (or disable the linter on a case-by-case basis):

```tsx
const foo = <T extends {}>(x: T) => x;
```

Which isn't great. See https://stackoverflow.com/questions/32308370/what-is-the-syntax-for-typescript-arrow-functions-with-generics

Similarly to the fact that function expressions that contain a `this` reference are allowed and will not create a failure, it would be nice if function expressions that have a generic type were allowed.

Thoughts?

**Comment by JoshuaKGoldberg** [comment]
> The following is not:
> ```typescript
> const foo = <T>(x: T) => x; // ERROR : unclosed `T` tag
> ```

Are you on the latest TypeScript? That works fine for me on typescriptlang.org/play ([link](http://www.typescriptlang.org/play/#src=const%20foo%20%3D%20%3CT%3E(x%3A%20T)%20%3D%3E%20x%3B%20%2F%2F%20ERROR%20%3A%20unclosed%20%60T%60%20tag%0D%0A%0D%0Afoo%3Cnumber%3E(3)%3B%0D%0Afoo%3Cnumber%3E(%22three%22)%3B)). Seems like that SO answer is outdated now?

**Comment by astorije** [comment]
Yes, latest TypeScript, but I believe this is a limitation of TSX. So the exclusion could be generic function in a TSX file.

**Comment by JoshuaKGoldberg** [comment]
> TSX

Ah! I'd missed that. Yes, agreed, this is a bug. +1 to excluding these generic functions in TSX files.
